### PR TITLE
Fix incoherent behaviors regarding user credentials management

### PR DESCRIPTION
**Pull request by scottinet**
# Description

* The API now forbids to create or update credentials on non-existing users
* Several fixes on credentials plugin calls in the security controllers (wrong strategy argument passed)
* The `reset` CLI command now thoroughly deletes users and their credentials, preventing inconsistent data when a kuzzle reset is performed
* The `resetStorage` function performing a Kuzzle reset has been moved from the main `kuzzle.js` file to the `cleanDb` CLI action implementation
* Kuzzle now throws internal errors when credentials are found on a non-existing user when attempting to create it

# Other changes

CLI commands now provide feedbacks to their caller (kuzzle or a CLI user) instead of writing them exclusively in Kuzzle logs.
This has been done primarily to make CLI users wait when performing a `reset`, as deleting users may take minutes depending on the amount of users to delete, and feedback is necessary to indicate that the CLI is performing actions (prior to this, the command would provide no feedback and users may think that the CLI has frozen)
CLI feedbacks have been extended to the `dump` and `shutdown` CLI commands.

# Fixed issue

#861

### CODE REVIEW

File: lib/api/controllers/cli/cleanDb.js [hunk 1]
```diff
@@ -19,6 +19,94 @@
...
+ * @returns Promise
+ */
+function resetStorage (kuzzle, statusUpdate) {
+  statusUpdate('Kuzzle reset initiated: this may take a while...');
+  
+  return deleteUsers(kuzzle, statusUpdate)
+    .then(() => kuzzle.internalEngine.deleteIndex())
+    .then(() => statusUpdate('Kuzzle internal database deleted'))
+    .then(() => {
+      const promises = ['internalCache', 'memoryStorage']
```
Comment (dbengsch) [inline lib/api/controllers/cli/cleanDb.js]: I would be more confortable if we only flushed the `internalCache` as we do for ES data, `memoryStorage` should not be flushed.

File: lib/api/controllers/securityController.js [hunk 2]
```diff
@@ -711,12 +690,25 @@ class SecurityController {
...
-      updateMethod = this.kuzzle.pluginsManager.getStrategyMethod(request.input.args.strategy, 'update'),
-      validateMethod = this.kuzzle.pluginsManager.getStrategyMethod(request.input.args.strategy, 'validate');
+    const 
+      id = request.input.resource._id,
+      strategy = request.input.args.strategy;
+
+    return this.kuzzle.repositories.user.load(id)
+      .then(user => {
+        if (user === null) {
+          throw new BadRequestError(`Cannot create credentials: unknown kuid "${id}"`);
```
Comment (dbengsch) [inline lib/api/controllers/securityController.js]: `update` instead of `create`

File: lib/api/controllers/cli/cleanDb.js (hunk 1 above)
Comment (scottinet) [inline lib/api/controllers/cli/cleanDb.js]: This is legacy code, I only moved this function without looking what it does. But I agree, clients storage shouldn't be resetted, at least until we add more options to the `reset` CLI action.

I'll remove the memory storage reset ASAP.

File: lib/api/controllers/securityController.js (hunk 2 above)
Comment (scottinet) [inline lib/api/controllers/securityController.js]: bart

**Pull request merged**